### PR TITLE
various documentation updates

### LINKED ISSUE #1352: documentations for datafusion-cli can be consolidated a bit more
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

currently there are two sections addressing usage in CLI, one in markdown and one in ast - i think we should consolidate them
Screen Shot 2021-11-23 at 14 25 53

**Describe the solution you'd like**

have one section to address these information

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #1813: Automatically test doc examples in user guide
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

While preparing for the 7.0.0 release, I manually stumbled on to the fact that the example on https://arrow.apache.org/datafusion/user-guide/example-usage.html (and linked in the main crate's readme) didn't work! see https://github.com/apache/arrow-datafusion/pull/1812

**Describe the solution you'd like**
I would like the examples in the user guide to be run as part of the normal CI in datafusion so that we have confidence they will work even if we do change the API

**Describe alternatives you've considered**
N/A

**Additional context**
@returnString  had this working once upon a time using the `doc-comments` crate -- https://github.com/apache/arrow/pull/9749 not sure what happened

### LINKED ISSUE #2558: Publish 8.0.0 user guide
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
We should publish the updated user guide based on the `8.0.0` release tarball (https://dlcdn.apache.org/arrow/arrow-datafusion-8.0.0/)

**Describe the solution you'd like**
We should publish the updated user guide based on the `8.0.0` release tarball

**Describe alternatives you've considered**
None

**Additional context**

### LINKED ISSUE #2866: [Question] interested in helping with documentation
Hi all,
I want to help with the documentation and I can pull issues from here to work on, but I have questions about how docs should be organized for the python project.

Datafusion-python is now in datafusion-contrib, outside of Apache. However, it's documentation is in the datafusion project repo. This leads to a weird dependency where docs will be created in the main datafusion repo, for versions of the python project managed by datafusion-contrib/datafusion-python. 

For example:
* there's a mismatch between datafusion's [python doc](https://github.com/apache/arrow-datafusion/blob/master/docs/source/python/index.rst) and the same documentation in the datafusion-python repo ([readme](https://github.com/datafusion-contrib/datafusion-python/blob/main/README.md))
* there are warnings when building the datafusion project docs that require changes to datafusion-python to fix. this is awkward.

It would be better if the docs for the datafusion-python project could be managed and published from that project, for consistency and ease of updating and possibly versioning of the docs in the future. Do you agree? Any ideas how to accomplish that? I'm willing to put in the work if there's agreement how to proceed.

### LINKED ISSUE #3001: User guide is incorrect regarding using CLI to register CSV files using schema inference
**Describe the bug**

The user guide page at https://arrow.apache.org/datafusion/cli/index.html states that "It is necessary to provide schema information for CSV files since DataFusion does not automatically infer the schema when using SQL to query CSV files." but this is not true, as demonstrated below:

```
DataFusion CLI v10.0.0
❯ create external table a stored as csv with header row location '/tmp/a.csv';
0 rows in set. Query took 0.017 seconds.
❯ select * from a;
+---+---+---+---+
| a | b | c | d |
+---+---+---+---+
| 1 | 2 | 3 | 4 |
+---+---+---+---+
1 row in set. Query took 0.011 seconds.
❯ describe a;
+-------------+-----------+-------------+
| column_name | data_type | is_nullable |
+-------------+-----------+-------------+
| a           | Int64     | NO          |
| b           | Int64     | NO          |
| c           | Int64     | NO          |
| d           | Int64     | NO          |
+-------------+-----------+-------------+
4 rows in set. Query took 0.017 seconds.
```

**To Reproduce**
See above.

**Expected behavior**
We should update the user guide to state that specifying a schema is optional.

**Additional context**
None

### PR DESCRIPTION
**Pull request by kmitchener**
# Which issue does this PR close?

Closes #1352
Closes #3001 
Closes #2866
You can also close #2558 and close #1813 though they're not directly related to this PR

 # Rationale for this change

# What changes are included in this PR?

* removes the python docs per issue #2866 (there's a related PR in the new python repo to bring the docs over there https://github.com/apache/arrow-datafusion-python/pull/23
* fix some links and typos and do consolidation of duplicate material
* replace some .rst files with .md in 
* fix docker build for datafusion-cli and update docs related to that
* improve left nav link naming for clarity (under Specification)
* consolidated the documentation for the CLI into one page per issue #1352
* fix the CSV schema inference in datafusion-cli docs per #3001
* update README.md to point to new location of DataFusion Python project
* update README.md with link to slides from Andrew Lamb and Daniel Heres

# Are there any user-facing changes?

The docs are slightly better.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +106,7 @@ There is no formal document describing DataFusion's architecture yet, but the fo
 
 - (March 2021): The DataFusion architecture is described in _Query Engine Design and the Rust-Based DataFusion in Apache Arrow_: [recording](https://www.youtube.com/watch?v=K6eCAVEk4kU) (DataFusion content starts [~ 15 minutes in](https://www.youtube.com/watch?v=K6eCAVEk4kU&t=875s)) and [slides](https://www.slideshare.net/influxdata/influxdb-iox-tech-talks-query-engine-design-and-the-rustbased-datafusion-in-apache-arrow-244161934)
 - (February 2021): How DataFusion is used within the Ballista Project is described in \*Ballista: Distributed Compute with Rust and Apache Arrow: [recording](https://www.youtube.com/watch?v=ZZHQaOap9pQ)
+- (May 2022): DataFusion and Apache Arrow: Supercharge Your Data Analytical Tool with a Rusty Query Engine [slides](https://docs.google.com/presentation/d/1q1bPibvu64k2b7LPi7Yyb0k3gA1BiUYiUbEklqW1Ckc/view#slide=id.g11054eeab4c_0_1165)
```
Comment (andygrove) [inline README.md:109]: This was just added in https://github.com/apache/arrow-datafusion/pull/3025

File: datafusion/core/src/lib.rs [hunk 2]
```diff
@@ -106,7 +106,7 @@
 //! Specifically, when DataFusion receives an SQL query, there are different steps
 //! that it passes through until a result is obtained. Broadly, they are:
 //!
-//! 1. The string is parsed to an Abstract syntax tree (AST) using [sqlparser](https://docs.rs/sqlparser/0.6.1/sqlparser/).
+//! 1. The string is parsed to an Abstract syntax tree (AST) using [sqlparser](https://docs.rs/sqlparser/0.18.0/sqlparser/).
```
Comment (andygrove) [inline datafusion/core/src/lib.rs:109]: is there a way we can just link to the latest version?

File: docs/source/user-guide/library.md [hunk 3]
```diff
@@ -19,54 +19,16 @@
 
 # Using DataFusion as a library
 
-## Create a new project
```
Comment (andygrove) [inline docs/source/user-guide/library.md:22]: Is it intentional that this section is removed?

Comment (andygrove) [reply docs/source/user-guide/library.md:22]: nm, it does seem rather redundant

File: docs/source/user-guide/sql/ddl.md [hunk 4]
```diff
@@ -30,9 +90,43 @@ STORED AS PARQUET
...
+    STORED AS CSV
+    WITH HEADER ROW
+    PARTITIONED BY (year)
+    LOCATION 'c:/tmp/data';
+```
+
+The above statement looks for CSV files in the `c:/tmp/data` directory and creates a table with
+the columns and data types inferred, as well as adding a column for the partition:
+
+TODO: describe rules for inference. which files does it look at, how many rows? is it configurable?
```
Comment (andygrove) [inline docs/source/user-guide/sql/ddl.md:113]: Currently the default number of rows to scan is defined here:

```
/// default max records to scan to infer the schema
pub const DEFAULT_SCHEMA_INFER_MAX_RECORD: usize = 1000;
```

This is not configurable via the DDL as far as I know.

File: README.md (hunk 1 above)
Comment (kmitchener) [reply README.md]: ok, updated PR and also added link to the slides

File: datafusion/core/src/lib.rs (hunk 2 above)
Comment (kmitchener) [reply datafusion/core/src/lib.rs]: done

**Comment by kmitchener** [comment]
Updated per review. I'm not done with this. Still working on organizing the content and creating new content.

**Comment by kmitchener** [comment]
@andygrove sure, yes, I can do that. I got a little exploratory and just kept making changes, unsure how they would work out. I'll submit separate PR for the simple cleanups and see how I can break this up further for easier review. I've been away from this for a few days, maybe fresh eyes will help.

**Comment by kmitchener** [comment]
closing this PR as it was exploratory and too big and all it's changes are going in other PRs

**Pull request closed without merging**